### PR TITLE
Support structured array, recarray, or just ndarray as mixin table columns

### PR DESCRIPTION
**Pull request by taldcroft**
This is re-do of #3759 using a mixin class `NdarrayMixin`.  This is a light subclass of `ndarray` that adds an `info` attribute and the minimum other machinery to function as a mixin column.

In addition to being able to explicitly view any `ndarray` as `NdarrayMixin` for table ingest, if an ndarray is a structured array or recarray then `Table` will automatically take the view and add to the table as a mixin.

cc: @mdboom @mhvk

### CODE REVIEW

File: astropy/table/table.py [hunk 1]
```diff
@@ -2211,3 +2219,45 @@ def _convert_col_for_table(self, col):
...
+        if six.callable(super(NdarrayMixin, self).__array_finalize__):
+            super(NdarrayMixin, self).__array_finalize__(obj)
+
+        # Self was created from template (e.g. obj[slice] or (obj * 2))
+        # or viewcast e.g. obj.view(Column).  In either case we want to
+        # init Column attributes for self from obj if possible.
+        if 'info' in getattr(obj, '__dict__', ()):
+            self.info = obj.info
+
+    def __reduce__(self):
```
Comment (taldcroft) [inline astropy/table/table.py]: I don't understand this fully but just lifted it from Quantity.

Comment (embray) [inline astropy/table/table.py]: Just a wild guess (though the link below might explain it better), it seems to me the ndarray `__setstate__` probably falls over if we pass in additional state variables that it doesn't know about.  So we have to separate the `ndarray` specific state variables from `ndarray.__reduce__` from the ones specific to our subclass, and then reconstitute them separately in `__setstate__`.

File: astropy/table/table.py [hunk 2]
```diff
@@ -436,6 +436,10 @@ def _init_from_list(self, data, names, dtype, n_cols, copy):
         def_names = _auto_names(n_cols)
 
         for col, name, def_name, dtype in zip(data, names, def_names, dtype):
+            # Structured ndarray gets viewed as a mixin
+            if isinstance(col, np.ndarray) and len(col.dtype) > 1:
+                col = col.view(NdarrayMixin)
```
Comment (embray) [inline astropy/table/table.py]: It feels a bit odd to me that plain ndarrays only use the `NdarrayMixin` if they are multi-field.  I realize it's unnecessary, but for the sake of consistency (as long as we're going wild with column mixins) would it hurt to use `NdarrayMixin` for all plain ndarray columns (i.e. the default)?

I'm just curious what the effect would be.

**Comment by embray** [comment]
I wish I understood better why this works as it does (I haven't been following the column mixin stuff as closely as I would like).  But it's awesome that this works! :fireworks:

**Comment by mhvk** [comment]
@embray [wrote] (https://github.com/astropy/astropy/pull/3925/files#r34155224):
> It feels a bit odd to me that plain ndarrays only use the `NdarrayMixin` if they are multi-field. I realize it's unnecessary, but for the sake of consistency (as long as we're going wild with column mixins) would it hurt to use `NdarrayMixin` for all plain ndarray columns (i.e. the default)?

For `Table`, the plain ndarray case is, of course, covered by `Column`. But I think it might it be an idea to follow this suggestion for `QTable`; if one were to just add `info` to any `MaskedArray`, one would also not need `MaskedColumn` any more, i.e., `QTable` would simply become a container for iterable items with the same length, to all of which it adds relevant `info`. 

@taldcroft - is this something worth considering a bit more? Or are there obvious downsides? (If so, as a separate issue; I don't want to distract from this one.)

File: astropy/table/tests/test_mixin.py [hunk 3]
```diff
@@ -79,13 +84,12 @@ def test_make_table(table_types, mixin_cols):
...
 
     cols = list(mixin_cols.values())
-    t = table_types.Table(cols, names=('a', 'b', 'c', 'm'))
+    t = table_types.Table(cols, names=('i', 'a', 'b', 'm'))
     check_mixin_type(t, t['m'], mixin_cols['m'])
 
     t = table_types.Table(cols)
     check_mixin_type(t, t['col3'], mixin_cols['m'])
 
-
```
Comment (mhvk) [inline astropy/table/tests/test_mixin.py]: Talk about trivial: two lines in between was correct PEP-wise.

File: astropy/utils/data_info.py [hunk 4]
```diff
@@ -384,3 +384,8 @@ def iter_str_vals(self):
 
 class MixinInfo(BaseColumnInfo):
     pass
+
+class ParentDtypeInfo(MixinInfo):
```
Comment (mhvk) [inline astropy/utils/data_info.py]: I think I'd make `QuantityInfo` a subclass of this one too (even though its definition would not change), just so a suitable `isinstance(info, ParentDtypeInfo)` would give the right answer.

**Comment by mhvk** [comment]
Now looked through the whole PR; looks good, modulo my one trivial and one really minor comment.

**Comment by taldcroft** [comment]
About the idea of moving toward using mixin columns more widely, including cases now handled by `Column` or `MaskedColumn`.  Interesting and it's worth some thought in long-term planning.  But in the current state of affairs we don't want to do this, because mixin columns are not required / expected to implement all behaviors of Column.  See [Mixin caveats](http://astropy.readthedocs.org/en/latest/table/mixin_columns.html#details-and-caveats) for details of what doesn't work.  Filling in those gaps is certainly a goal.

About MaskedColumn in particular, that has gotten a lot of TLC to make MaskedArrays actually work well within astropy (most of the time) by patching over MaskedArray bugs / "features".

**Comment by mhvk** [comment]
@taldcroft - OK, somewhat further in the future then...

**Pull request merged**